### ISSUE TITLE
The documentation is not clear on how to use cucumber js to test cli applications

### ISSUE THREAD

**Issue by dklymenk** [cucumber/cucumber-js#1865]
**Is your feature request related to a problem? Please describe.**
I am trying to use cucumber js to drive the development of my cli tool. But I am unable to find any documentation or online article on it.

**Describe the solution you'd like**
Perhaps a documentation page on your website similar to the one you have for [Browser Automation](https://cucumber.io/docs/guides/browser-automation/) would work the best.

**Describe alternatives you've considered**
1. Using node child processes to run the command with different arguments and flags.
This approach would work ok for a simple input/output command, but since the cli utility I am working on will be sending actual requests to both GET and POST data on remote server, I would prefer to have an ability to mock the requests and run expectations on the data that would be posted to the server for example.
2. I am going to use [oclif](https://oclif.io/) as a framework for my app and it comes with pretty poorly documented, yet functional set of [testing utilities](https://oclif.io/docs/testing). I tried using them in my cucumber step definitions file, however when I try do `import {test} from '@oclif/test'` I get an error `ReferenceError: it is not defined`. It appears that the test suite for that framework requires `mocha` to be used a test runner, which we can't really do with cucumber js.  

**Additional context**
Basically I am mostly curious on what's the supported way to use cucumber js for writing step definitions for a command line application. I fine using another library/framework for this app if it means going the intended route without hacking my way though it by relying on unsupported or sub-optimal solutions.

**Comment by aurelien-reeves** [comment]
Hi :)

Actually, cucumber-js, which is a CLI tool itself, is testing itself :D

Did you had the opportunity to look at cucumber-js own tests? You may find the following interesting:
- [A feature dedicated to the CLI of cucumber-js](https://github.com/cucumber/cucumber-js/blob/main/features/cli.feature.md)
- [The world used with step definitions and which define some CLI interactions](https://github.com/cucumber/cucumber-js/blob/main/features/support/world.ts)

Regarding your project, do you need to consider your CLI tool as a black box? Otherwise, you can also test mainly the internals directly, and only a few test to make sure the CLI itself is working well

**Comment by dklymenk** [comment]
Thanks for your swift reply.

> Regarding your project, do you need to consider your CLI tool as a black box? Otherwise, you can also test mainly the internals directly, and only a few test to make sure the CLI itself is working well

The CLI tool will simply parse the input csv file and send HTTP requests to another server. So with cucumber I want to make sure  it sends requests that reflect the contents of the input csv. That is the main functionality at least. I would also want to make sure I get some error if csv is invalid, but other than that I don't really care about stdout either. So with my testing strategy I would like to treat it as a black box as much possible. The features would look similar to the following, but with more specifics of course:
```
Given I have a csv file with contents like this
And I have a configuration file like this
And a GET request to this route returns this JSON
When I run the command
Then a POST request is sent with a JSON body like this
```

While the cli feature you linked looks beautiful, the code behind it looks very intimidating. Especially the `run` methods on `World` and `Cli` classes, but as far as I can tell it doesn't actually do anything complicated, and since when using the node js frameworks (like [commander.js](https://github.com/tj/commander.js) or [oclif](https://oclif.io/)) for writing commands you can already run them programatically, I won't need that much code. The biggest trick to me so far seems to be abstracting the api interactions in such a way that if for example `APP_ENV` is `test` it would lead to mocks I describe with step definitions.

I will give a few more things a try and come back if I have came up with something worthwhile.

Thanks again.

**Comment by aurelien-reeves** [comment]
>
> While the cli feature you linked looks beautiful, the code behind it looks very intimidating. Especially the `run` methods on `World` and `Cli` classes, but as far as I can tell it doesn't actually do anything complicated, and since when using the node js frameworks (like [commander.js](https://github.com/tj/commander.js) or [oclif](https://oclif.io/)) for writing commands you can already run them programatically, I won't need that much code. The biggest trick to me so far seems to be abstracting the api interactions in such a way that if for example `APP_ENV` is `test` it would lead to mocks I describe with step definitions.
> 

I would avoid to instrument the code in order to execute some tests.
In your case considering the subject under test as a black box, I guess I would try to rely on sinon-js to fake a server, or even to write a dedicated test server rather than instrumenting the production code to detect a testing environment.

> 
> Thanks again.

You are very welcome :)

**Comment by dklymenk** [comment]
Ok, so I have spent quite some time on playing around with this and can share some of my findings. While I am not very experienced in this field, I think anyone looking into using cucumber js for driving the development of their cli app would find something useful below.

A link to the repo: https://github.com/dklymenk/cucumber-oclif-jsonplaceholder-playground

I'll split the narrative in 3 parts for each of the ways an app would interact with the user or outside world: stdin, stdout and http.

## stdout

I went with https://github.com/jdxcode/stdout-stderr. It provides a simple API and had the most recent updates among other libraries that you get when searching for "node mock stdout". There is one major downside to using this library and it's the fact that it would siphon every single line of output which makes using console.log for debugging not just impossible, but also breaks other scenarios where your assertions start failing because of random debug output.

So if you rely on console.log for debugging I would suggest mocking methods you use for output and not entire stdout. If you're using `oclif` that means mocking `log`, `warn`, `error` methods of the base class. This means your feature tests will be coupled to implementation more than one might like, but that the compromise you'll need to take if you rely on console.log for debugging.

## stdin

I tried to use https://github.com/caitp/node-mock-stdin, but no matter how I plugged it in, I was still being prompted in the middle of the tests.

I ended up just stubbing the library method I use for prompts. I am not happy with it, but I just didn't find a better option.

## http

My initial instinct here was to use https://github.com/nock/nock , because it has a clean syntax and has support for everything I thought I needed, however it turned out to be very hard to write step definitions using that library because it forces you to write expectations at the same time you declare the mock. Meaning you can't declare a mock during a Given step and run expectations during a Then step. I didn't really want to declare my expectations during a Given step, so I started looking for an alternative.

Sinon's fake server (https://sinonjs.org/releases/v12.0.1/fake-xhr-and-server/) seemed like a good replacement since it allows to look at request history to make assertions. The issue is that it only works in browser environment since it interacts with `XMLHttpRequest` and doesn't touch node.js `http` and `https` modules at all.

So my only option again was to mock the library I use to send requests... Since I already had sinon installed, I thought I would use it for mocks too, but it also proved hard to use in a Given When Then format because you had to declare expectations before mocks are executed.

So for http "mocking" I ended up stubbing individual axios methods and then spying on the entire object with sinon.

## Summary

While I have learned a lot here. The main takeaway for me was that unless your client or business requires you to do BDD, you shouldn't use it for node cli apps. The available toolchain is simply not good enough. You have to write tighly coupled step definitions and write Gherkin in manner you wouldn't normally write in when describing behaviour of a web or mobile app.

Anyway, I would be happy if I were to be proven wrong. It would be great to see not just official documentation on BDD with cucumber-js for cli apps, but also improve this doc https://cucumber.io/docs/cucumber/mocking-and-stubbing-with-cucumber/#javascript with some examples and recommendations as I personally was struggling with mocks because I couldn't declare them at a Given step and run expectations on a Then step.

**Comment by aurelien-reeves** [comment]
Thanks @dklymenk for your post, and for the time you took to make all those experiments! This is great feedback!

Just a quick though: you are using axios for your http requests. Do you know that there are at least one adapter dedicated to mock that library: [axios-mock-adapter](https://www.npmjs.com/package/axios-mock-adapter) for example? Maybe that would be easier to work with something like that than by mocking it manually?

**Comment by dklymenk** [comment]
Yes, that particular one you linked actually has a [history](https://www.npmjs.com/package/axios-mock-adapter#history) property we can use to make assertions after the requests are made.

I will definitely try it out. Thank you.

Actually I don't have any strong opinions on [axios](https://github.com/axios/axios) vs [node-fetch](https://github.com/node-fetch/node-fetch), so if using node-fetch would open up additional options that better fit the Given When Then test structure, I personally won't mind using it.

**Comment by dklymenk** [comment]
Your suggestion to try out [axios-mock-adapter](https://www.npmjs.com/package/axios-mock-adapter) was great!

While the step definition now looks a little bit more messed up and requires a helper function to capitalise the first letter of method name, I like this approach much more. https://github.com/dklymenk/cucumber-oclif-jsonplaceholder-playground/commit/d357a811b94d63a1b78af17d060d301607e8a984

It also inspired me to look up requests in history by their url not just blindly getting first request, which wasn't a good long term solution in the first place. https://github.com/dklymenk/cucumber-oclif-jsonplaceholder-playground/commit/29451b26e3308b9ecfa85e2748ac83ca9e9fa800

Anyway, thank you very much. I am looking forward to checking the updated documentation once it arrives.

**Comment by aurelien-reeves** [comment]
Also, I see that you pay a great attention to the organization of your code and your step definitions. Maybe you could be interested into https://github.com/cucumber/screenplay.js

**Comment by dklymenk** [comment]
> Also, I see that you pay a great attention to the organization of your code and your step definitions. Maybe you could be interested into https://github.com/cucumber/screenplay.js

Looks great. I will consider it next time I start a full stack project. When it comes to CLI app that interacts with remote servers with set contract, there is only one assembly I can think of, so doesn't seem like the library is usable for such case.

**Comment by davidjgoss** [comment]
Looks like we got to a good place with this, going to close now and also note that our examples repo has a simple CLI example https://github.com/cucumber-examples/cucumber-js-examples. Thanks for all the feedback @dklymenk!
